### ISSUE TITLE
How to pytest aiohttp when used in FastAPI

### ISSUE THREAD

**Issue by writeson** [tiangolo/fastapi#5426]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import aiohttp

app = FastAPI()


@app.get("/")
async def root():
    async with aiohttp.ClientSession() as session:
        async with session.get("http://httpbin.org/get") as response:
            return response.json()

# I'm not sure how to set up a pytest function to mock and test apihttp
```

### Description

I've got a FastAPI application that makes HTTP GET requests to another service using the Requests module. I'd like to make these endpoints asynchronous using aiohttp, but I don't know how to build pytest functions to mock and test the functions with aiohttp in them.

### Operating System

Linux

### Operating System Details

This is running on a Docker container with Python 3.9.7

### FastAPI Version

0.84.0

### Python Version

3.9.7

**Comment by JarroVGIT** [comment]
Not really a FastAPI question per se, more of a general "how to mock aiohttp in pytest", no?

**Comment by humrochagf** [comment]
Yup FastAPI will not have a direct impact on how you test it, this should work the same for other frameworks that support async as well.

To mock `aiohttp` calls there's a lib do it that is called `aioresponses`.

You can check their repo at https://github.com/pnuckowski/aioresponses

The mocking of aiohttp doesn't impact any of the testing flows that you normally have when testing fastapi, you can even use `TestClient` (that has a sync API) to write your test like this:

```python
import aiohttp
import pytest
from aioresponses import aioresponses
from fastapi import FastAPI
from fastapi.testclient import TestClient

app = FastAPI()


@app.get("/")
async def root():
    async with aiohttp.ClientSession() as session:
        async with session.get("http://httpbin.org/get") as response:
            return await response.json()


@pytest.fixture
def anyio_backend():
    return 'asyncio'


@pytest.mark.anyio
async def test_root():
    client = TestClient(app)

    with aioresponses() as mocked:
        mocked.get("http://httpbin.org/get", status=200, body='{"test": true}')

        response = client.get("/")

        assert response.json() == {"test": True}
```

**Comment by jgould22** [comment]
If you are using pytest then you can use its monkeypatch function

**Comment by writeson** [comment]
>

**Comment by writeson** [comment]
@JarroVGIT 
Yes, clearly, I need to understand FastAPI and PyTest (as well as mocking) better! :)

**Comment by writeson** [comment]
I can still use some help writing a test that looks something like this:

```python
async def test_validation_token():
    response = await validate_token("test_token")
    assert response is True
```

Where validate_token() is a function in a module that uses aiohttp to make a "GET" request to another service.

**Comment by humrochagf** [comment]
The first parameter of `mocked.get` in my example is the path that you want to mock, so if the internal request that you make with `aiohttp` matches the path that you pass to this function then you have total control of it's returned values with the other parameters that you are passing to it like the `status` and the `body` of the response.

Just make sure you are calling your `validate_token` inside the context manager of the `aioresponses`

**Comment by writeson** [comment]
@humrochagf 
Thanks for the clarification. I'll give that a try and see how far I get. I'm constantly confused by mocking and if mocking in one function will affect a call to the object being mocked in another module. @patch seems clearer to me, but as I said, I'm a relative noob to pytest.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @writeson  👍
